corehq/util/couch_helpers.py

```diff
@@ -169,6 +171,153 @@ def paginate_view(db, view_name, chunk_size,
+    iteration checkpoint. In the worst case about this many documents
+    that were previously yielded may be yielded again if the iteration
+    is stopped and later resumed.
+    """
+
+    def __init__(self, db, doc_types, iteration_key, chunk_size=100):
+        if isinstance(doc_types, str):
+            raise TypeError("expected list of strings, got %r" % (doc_types,))
+        self.db = db
+        self.original_doc_types = doc_types = sorted(doc_types)
+        self.iteration_key = iteration_key
+        self.chunk_size = chunk_size
+        iteration_name = "{}/{}".format(iteration_key, " ".join(doc_types))
+        self.iteration_id = hashlib.sha1(iteration_name).hexdigest()
+        try:
+            self.state = db.get(self.iteration_id)
+        except ResourceNotFound:
+            # new iteration
+            self.state = {
+                "_id": self.iteration_id,
+                "doc_type": "ResumableDocsByTypeIteratorState",
+                "retry": {},
+
+                # for humans
+                "name": iteration_name,
+                "timestamp": datetime.datetime.utcnow().isoformat()
+            }
+            args = {}
+        else:
+            # resume iteration
+            args = self.state.get("offset", {}).copy()
+            if args:
+                assert args.get("startkey"), args
+                doc_type = args["startkey"][0]
+                # skip doc types before offset
+                doc_types = doc_types[doc_types.index(doc_type):]
+            else:
+                # non-retry phase of iteration is complete
+                doc_types = []
+        args.update(
+            view_name='all_docs/by_doc_type',
+            chunk_size=chunk_size,
+            log_handler=ResumableDocsByTypeLogHandler(self),
+            include_docs=True,
+            reduce=False,
+        )
+        self.view_args = args
+        self.doc_types = doc_types
+
+    def __iter__(self):
+        args = self.view_args
+        for doc_type in self.doc_types:
+            if args.get("startkey", [None])[0] != doc_type:
+                args.pop("startkey_docid", None)
+                args["startkey"] = [doc_type]
+            args["endkey"] = [doc_type, {}]
+            for result in paginate_view(self.db, **args):
+                yield result['doc']
+
+        retried = {}
+        while self.state["retry"] != retried:
+            for doc_id, retries in list(self.state["retry"].iteritems()):
+                if retries == retried.get(doc_id):
+                    continue  # skip already retried (successfully)
+                retried[doc_id] = retries
+                try:
+                    yield self.db.get(doc_id)
+                except ResourceNotFound:
+                    pass
+
+        # save iteration state without offset to signal completion
+        self.state.pop("offset", None)
+        self.state["retry"] = {}
+        self._save_state()
```

Comment:
note to self: should save state without offset before retry logic, and only save again if any docs were retried.